11_FFTBloom/app_resources/fft_convolve_ifft.hlsl

```diff
@@ -0,0 +1,205 @@
+#include "common.hlsl"
+
+/*
+ * Remember we have these defines:
+ * _NBL_HLSL_WORKGROUP_SIZE_
+ * ELEMENTS_PER_THREAD
+ * KERNEL_SCALE
+ * scalar_t
+ * FORMAT
+*/
+
+// Can't specify format
+/* [[vk::combinedImageSampler]]*/ [[vk::binding(0, 0)]] /* [[vk::image_format(FORMAT)]]*/ Texture2D<float32_t2> kernelChannels[CHANNELS] : register(t0);
+/* [[vk::combinedImageSampler]]*/ [[vk::binding(0, 0)]] SamplerState samplerState[CHANNELS] : register(s0);
+
+#define FFT_LENGTH (_NBL_HLSL_WORKGROUP_SIZE_ * ELEMENTS_PER_THREAD)
+
+groupshared uint32_t sharedmem[workgroup::fft::SharedMemoryDWORDs<scalar_t, _NBL_HLSL_WORKGROUP_SIZE_>];
+
+// Users MUST define this method for FFT to work
+uint32_t3 glsl::gl_WorkGroupSize() { return uint32_t3(_NBL_HLSL_WORKGROUP_SIZE_, 1, 1); }
+
+struct SharedMemoryAccessor
+{
+	void set(uint32_t idx, uint32_t value)
+	{
+		sharedmem[idx] = value;
+	}
+
+	void get(uint32_t idx, NBL_REF_ARG(uint32_t) value)
+	{
+		value = sharedmem[idx];
+	}
+
+	void workgroupExecutionAndMemoryBarrier()
+	{
+		glsl::barrier();
+	}
+
+};
+
+// ---------------------------------------------------- Utils ---------------------------------------------------------
+uint64_t colMajorOffset(uint32_t x, uint32_t y)
+{
+	return x * glsl::gl_NumWorkGroups().x | y; // can sum with | here because NumWorkGroups is still PoT (has to match half the FFT_LENGTH of previous pass)
+}
+
+uint64_t rowMajorOffset(uint32_t x, uint32_t y)
+{
+	return y * pushConstants.dataElementCount + x; // can no longer sum with | since there's no guarantees on row length
+}
+
+// Same as what was used to store in col-major after first axis FFT. This time we launch one workgroup per row so the height of the channel's (half) image is `glsl::gl_NumWorkGroups().x`,
+// and the width (number of columns) is passed as a push constant
+uint64_t getColMajorChannelStartAddress(uint32_t channel)
+{
+	return pushConstants.colMajorBufferAddress + channel * glsl::gl_NumWorkGroups().x * pushConstants.dataElementCount * sizeof(complex_t<scalar_t>);
+}
+
+// Image saved has the same size as image read 
+uint64_t getRowMajorChannelStartAddress(uint32_t channel)
+{
+	return pushConstants.rowMajorBufferAddress + channel * glsl::gl_NumWorkGroups().x * pushConstants.dataElementCount * sizeof(complex_t<scalar_t>);
+}
```

Comment:
why not make them members of the `push constants` struct ?

Reply:
They all differ between shaders, it's doable though